crates/holochain/src/conductor/conductor/agent_key_operations.rs

```diff
@@ -0,0 +1,152 @@
+
+        // Enable app again.
+        self.clone().enable_app(app_id.clone()).await?;
+
+        let deletion_per_cell_results = revocation_per_cell_results?;
+
+        // Publish 'Delete' actions of cells where successful.
+        // Triggering workflow is only possible when cells are enabled.
+        let publish_workflow_triggers = deletion_per_cell_results
+            .iter()
+            .filter(|(_, result)| result.is_ok())
+            .map({
+                |(cell_id, _)| {
+                    let conductor = self.clone();
+                    async move {
+                        let cell_triggers = conductor.get_cell_triggers(cell_id).await?;
+                        cell_triggers.publish_dht_ops.trigger(&"delete agent key");
+                        ConductorApiResult::Ok(())
+                    }
+                }
+            });
+        let _trigger_publish_results = futures::future::join_all(publish_workflow_triggers).await;
+
+        // Return cell ids with their agent key deletion result
+        Ok(deletion_per_cell_results)
+    }
+
+    /// Revoke agent key in Deepkey first, if installed, and then write a [`Delete`] of the key to the source chain.
+    async fn revoke_agent_key_for_app_inner(
+        conductor: Arc<Conductor>,
+        agent_key: AgentPubKey,
+        app_id: InstalledAppId,
+    ) -> ConductorResult<RevokeAgentKeyForAppResult> {
+        // If DPKI service is installed, revoke agent key there first
+        if let Some(dpki_service) = conductor.running_services().dpki {
+            let dpki_state = dpki_service.state().await;
+            let timestamp = Timestamp::now();
+            let key_state = dpki_state.key_state(agent_key.clone(), timestamp).await?;
+            match key_state {
+                KeyState::NotFound => {
+                    return Err(ConductorError::DpkiError(
+                        DpkiServiceError::DpkiAgentMissing(agent_key.clone()),
+                    ))
+                }
+                // If the key already is invalid, do nothing. Operation should be idempotent to allow for
+                // retries if agent key of some source chain could not be deleted successfully.
+                KeyState::Invalid(_) => (),
+                KeyState::Valid(_) => {
+                    // Get action hash of key registration
+                    let key_meta = dpki_state.query_key_meta(agent_key.clone()).await?;
+                    // Sign revocation request
+                    let signature = dpki_service
+                        .cell_id
+                        .agent_pubkey()
+                        .sign_raw(
+                            &conductor.keystore,
+                            key_meta.key_registration_addr.get_raw_39().into(),
+                        )
+                        .await
+                        .map_err(|e| DpkiServiceError::Lair(e.into()))?;
+                    let signature = deepkey_roundtrip_backward!(Signature, &signature);
```

Comment:
@maackle ?

Reply:
Boilerplate reduction. Two things that only a macro can do:
- Importing two different types of the same name from two different crates (and only needing to specify the type name once)
- Handling two unwraps inline, so that if there is a panic, we'll see the line of code that it happened (instead of a function body)